roles/adoption_osp_deploy/tasks/main.yml

```diff
@@ -0,0 +1,140 @@
+# All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+- name: Ensure controller and undercloud vms are started and reachable
+  vars:
+    _cifmw_libvirt_manager_layout:
+      vms:
+        controllers:
+          start: true
+          disk_file_name: "dummy"
+        osp-underclouds:
+          start: true
+          disk_file_name: "dummy"
+    cifmw_libvirt_manager_all_vms: >-
+      {%- set vms = {} -%}
+      {%- for vm in _vm_groups['osp-underclouds'] -%}
+        {%- set _ = vms.update({vm: 'osp-underclouds'}) -%}
+      {%- endfor -%}
+      {%- for vm in _vm_groups['controllers'] -%}
+        {%- set _ = vms.update({vm: 'controllers'}) -%}
+      {%- endfor -%}
+      {{ vms }}
+    ansible_libvirt_pools: {}
+  ansible.builtin.include_role:
+    name: "libvirt_manager"
+    tasks_from: "start_vms.yml"
+    apply:
+      delegate_to: "{{ cifmw_target_host | default('localhost') }}"
+
+- name: Gather ansible_user_id from undercloud
+  delegate_to: "osp-undercloud-0"
+  ansible.builtin.setup:
+    gather_subset:
+      - user_id
+      - user_dir
+
+- name: Generate configuration files
+  ansible.builtin.include_tasks: config_files.yml
+  vars:
+    stack: "{{ overcloud_vars }}"
+  loop: "{{ _adoption_source_scenario.stacks }}"
+  loop_control:
+    loop_var: overcloud_vars
+
+- name: Prepare undercloud enviornment
+  tags:
+    - undercloud
+  ansible.builtin.import_tasks: prepare_undercloud.yml
+
+- name: Stop before running pre undercloud hooks
+  tags:
+    - undercloud
+  vars:
+    _stage_stopper: "before_pre_hook_undercloud"
+  when:
+    - cifmw_adoption_osp_deploy_stopper is defined
+    - cifmw_adoption_osp_deploy_stopper in _stage_stopper
+  ansible.builtin.fail:
+    msg: "Failing on demand {{ cifmw_adoption_osp_deploy_stopper }}"
+
+- name: Run pre undercloud deploy hooks
+  tags:
+    - undercloud
+  vars:
+    hooks: "{{ _adoption_source_scenario.pre_uc_run | default([]) }}"
+    step: "pre_undercloud"
+  ansible.builtin.include_role:
+    name: run_hook
+
+- name: Stop before deploying undercloud
+  tags:
+    - undercloud
+  vars:
+    _stage_stopper: "before_deploy_undercloud"
+  when:
+    - cifmw_adoption_osp_deploy_stopper is defined
+    - cifmw_adoption_osp_deploy_stopper in _stage_stopper
+  ansible.builtin.fail:
+    msg: "Failing on demand {{ cifmw_adoption_osp_deploy_stopper }}"
+
+- name: Deploy undercloud
+  tags:
+    - undercloud
+  ansible.builtin.import_tasks: deploy_undercloud.yml
```

Comment:
Question: I note this task file has a single task in it, are you planning to call it from other places? Is there a reason you extracted it to it's own file?

Reply:
No, I don't expect to call it anywhere else. When I started writing the role it had more tasks, but then I refactored some of them out and it was left with one task. I'll move the task to the main file, since it's quite small